### PR TITLE
do not mark lightningmodule methods as abstract

### LINKED ISSUE #8752: Make default implementation for train_dataloader in DataHooks raise NotImplementedError
### Discussed in https://github.com/PyTorchLightning/pytorch-lightning/discussions/8734

We are auditing the Lightning components and APIs to assess opportunities for improvements:
- https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

The default implementation for these is logging a warning that nothing is implemented. https://github.com/PyTorchLightning/pytorch-lightning/blob/963c26764682fa4cf64c93c5a7572ae0040e9c32/pytorch_lightning/core/hooks.py#L529

The current default implementation logs a warning, which can be lost, and which forces the framework to perform is_overridden checks. Instead, having the default implementation raise a NotImplementedError will make errors clear to users in case users forget to implement these hooks.

This applies for all the dataloader hooks in DataHooks, not just train_dataloader

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #8752
Follow up to #9499 (not blocking each other)

Recently, #8752 added NotImplementedError being raised in LightningModule data hooks, but this marks the methods as abstract and linters complain that user did not implement all abstract methods. 
This is confusing and a bad user experience, since our docs never state that you need to implement these methods. In fact, we always recommend that a datamodule should be considered first!

image

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

**Review by SeanNaren — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]

File: pytorch_lightning/core/hooks.py [hunk 1]
```diff
@@ -471,7 +472,7 @@ def train_dataloader(self):
                 # each batch will be a dict of tensors: {'mnist': batch_mnist, 'cifar': batch_cifar}
                 return {'mnist': mnist_loader, 'cifar': cifar_loader}
         """
-        raise NotImplementedError("`train_dataloader` must be implemented to be used with the Lightning Trainer")
```
Comment (carmocca) [inline pytorch_lightning/core/hooks.py:474]: I don't get why PyCharm shows that warning. Other classes in our codebase raise `NotImplementedError` but they don't have that warning. For example:

https://github.com/PyTorchLightning/pytorch-lightning/blob/23450e29055addcf8f1100b4fc3a4ed4fa9c0e2d/pytorch_lightning/overrides/base.py#L39

https://github.com/PyTorchLightning/pytorch-lightning/blob/23450e29055addcf8f1100b4fc3a4ed4fa9c0e2d/pytorch_lightning/callbacks/progress/base.py#L140

(among others)

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/core/hooks.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/core/hooks.py:474]: Yes, I also find it rude. If this changes in the future, we can change it back too.

Comment (carmocca) [reply pytorch_lightning/core/hooks.py:474]: minimal repro:

```python
class Foo:
    def fn(self):
        raise NotImplementedError


class HasWarning(Foo):
    ...


class NoWarning:
    def fn(self):
        raise NotImplementedError
```

This looks like a false positive. Already reported in https://youtrack.jetbrains.com/issue/PY-38680

We can just disable the warning

Comment (awaelchli) [reply pytorch_lightning/core/hooks.py:474]: do you recommend to close the PR? is it important that we raise NotImplemented?
I read the [docs](https://docs.python.org/3/library/exceptions.html#NotImplementedError) and interpret it differently

Comment (carmocca) [reply pytorch_lightning/core/hooks.py:474]: We can close this or add `# noqa`

Unfortunately, there's no mechanism to silence that warning in particular so any other warnings would be hidden.

I personally lean towards closing this but no strong opinion.

Comment (awaelchli) [reply pytorch_lightning/core/hooks.py:474]: I tried adding noqa to the lines where we raise NotImiplementedError and the signature of the method but both did not work

Comment (carmocca) [reply pytorch_lightning/core/hooks.py:474]: `class LightningModule(  # noqa` works for me

No need to add it to the `NotImplementedError`

Comment (ananthsub) [reply pytorch_lightning/core/hooks.py:474]: If adding the noqa marker, could we link to the JetBrains issue in the code to know why it's there? 

Once that issue is fixed I'd prefer using NotImplementedError as its canonical Python while Misconfiguration is Lightning specific. It's very minor though, as long as we raise an exception compared to before of logging a warning message.

Comment (awaelchli) [reply pytorch_lightning/core/hooks.py]: @carmocca any adding noqa to the base class gets rid of the warning there but not in any of the derived classes.
